dali/imgcodec/util/convert.cc

```diff
@@ -0,0 +1,129 @@
+// limitations under the License.
+
+#include <cassert>
+#include <utility>
+#include "dali/core/static_switch.h"
+#include "dali/imgcodec/util/convert.h"
+#include "dali/kernels/slice/slice_cpu.h"
+#include "dali/kernels/common/utils.h"
+
+namespace dali {
+namespace imgcodec {
+
+template <int static_ndim = -1, typename Out, typename In, typename ConvertFunc>
+void Convert(Out *out, const int64_t *out_strides,
+             const In *in, const int64_t *in_strides,
+             const int64_t *size, int ndim,
+             ConvertFunc &&func) {
+  if constexpr (static_ndim < 0) {
+    VALUE_SWITCH(ndim, NDim, (0, 1, 2, 3, 4),
+      (Convert<NDim>(out, out_strides, in, in_strides, size, NDim,
+                     std::forward<ConvertFunc>(func));
+      return;), ()
+    );  // NOLINT
+  }
+
+  int64_t extent = size[0];
+  int64_t in_stride = in_strides[0];
+  int64_t out_stride = out_strides[0];
+
+  if constexpr (static_ndim == 0) {
+    func(out, in);
+  } else if constexpr (static_ndim == 1) {  // NOLINT - if constexpr not recognized
+    for (int64_t i = 0; i < extent; i++) {
+      func(out + i * out_stride, in + i * in_stride);
+    }
+  } else {
+    assert(ndim != 1 && "This should go with the static ndim codepath");
+    for (int64_t i = 0; i < extent; i++) {
+      const int next_ndim = static_ndim < 0 ? -1 : static_ndim - 1;
+      Convert<next_ndim>(out + i * out_stride, out_strides + 1,
+                         in + i * in_stride, in_strides + 1,
+                         size + 1, ndim - 1,
+                         std::forward<ConvertFunc>(func));
+    }
+  }
+}
+
+template <typename Out, typename In>
+void ConvertDType(Out *out, const In *in) {
+  *out = ConvertSatNorm<Out>(*in);
+}
+
+#define IMG_CONVERT_TYPES uint8_t, int8_t, uint16_t, int16_t, uint32_t, int32_t, float, float16
+
+void Convert(SampleView<CPUBackend> out, TensorLayout out_layout, DALIImageType out_format,
+             ConstSampleView<CPUBackend> in, TensorLayout in_layout, DALIImageType in_format,
+             TensorShape<> roi_start, TensorShape<> roi_end) {
+  TensorShape<> out_shape = out.shape();
+  TensorShape<> in_shape = in.shape();
+  assert(in_shape.sample_dim() == out_shape.sample_dim());
+  assert(in_shape.sample_dim() == in_shape.sample_dim());
+  int ndim = out_shape.sample_dim();
+  int spatial_ndim = ndim - 1;
+  DALI_ENFORCE(ImageLayoutInfo::ChannelDimIndex(in_layout) == ndim - 1,
+    "Not implemented: currently only channels-last layout is supported");
+  DALI_ENFORCE(in_layout == out_layout,
+    "Not implemented: currently layout transposition is not supported");
+
+  // NOTE: there's opencv-based color space conversion, but it may be unable
+  // to also perform data type conversion, which we want.
+  DALI_ENFORCE(out_format == in_format,
+    "Not implemented: color space conversion is not implemented");
+
+  if (!roi_start.empty() && roi_start.sample_dim() != spatial_ndim)
```

Comment:
We have `use_roi` field in `DecoderParams`, which you set, but never read in this PR. Instead, you rely on an assumption that `roi.begin()` and `roi.end()` are empty if ROI is not used. 

If you think that's better, maybe let's remove `use_roi` and adjust the docs for `DecoderParams` stating that `roi.begin()` and `roi.end()` are empty iff ROI is not used? Otherwise, I think we should rely on the `use_roi` field instead of those emptiness checks

Reply:
`use_roi` has been removed.